SIGSEGV: segmentation violation

When using gas to scan  golang.org/x/crypto/acme/autocert/autocert.go, I get the following segmentation violation:

```
[gas] 2018/02/13 16:08:38 Checking package: autocert
[gas] 2018/02/13 16:08:38 Checking file: /Users/browne/workspace/go/src/golang.org/x/crypto/acme/autocert/autocert.go
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x126a7ec]
goroutine 1 [running]:
github.com/GoASTScanner/gas/rules.(*insecureConfigTLS).Match(0xc42007d920, 0x1474fc0, 0xc420239ec0, 0xc420054400, 0x1, 0xdc0000c420c2cb48, 0xc42017a928)
	/Users/browne/go/src/github.com/GoASTScanner/gas/rules/tls.go:109 +0x7c
github.com/GoASTScanner/gas.(*Analyzer).Visit(0xc4200841e0, 0x1474fc0, 0xc420239ec0, 0x2, 0x14e2fe0)
	/Users/browne/go/src/github.com/GoASTScanner/gas/analyzer.go:171 +0x1c5
go/ast.Walk(0x1473980, 0xc4200841e0, 0x1474fc0, 0xc420239ec0)
	/usr/local/go/src/go/ast/walk.go:52 +0x66
go/ast.Walk(0x1473980, 0xc4200841e0, 0x1475940, 0xc420240e80)
	/usr/local/go/src/go/ast/walk.go:143 +0x15df
go/ast.walkExprList(0x1473980, 0xc4200841e0, 0xc4205a7e70, 0x1, 0x1)
	/usr/local/go/src/go/ast/walk.go:26 +0x81
go/ast.Walk(0x1473980, 0xc4200841e0, 0x1474c40, 0xc420239f00)
	/usr/local/go/src/go/ast/walk.go:207 +0x211f
go/ast.walkStmtList(0x1473980, 0xc4200841e0, 0xc420244000, 0x3, 0x4)
	/usr/local/go/src/go/ast/walk.go:32 +0x81
go/ast.Walk(0x1473980, 0xc4200841e0, 0x1474e80, 0xc420244040)
	/usr/local/go/src/go/ast/walk.go:238 +0x1e95
go/ast.walkStmtList(0x1473980, 0xc4200841e0, 0xc420244100, 0x3, 0x4)
	/usr/local/go/src/go/ast/walk.go:32 +0x81
go/ast.Walk(0x1473980, 0xc4200841e0, 0x1474dc0, 0xc42023b350)
	/usr/local/go/src/go/ast/walk.go:224 +0x1b71
go/ast.Walk(0x1473980, 0xc4200841e0, 0x1475900, 0xc42023b380)
	/usr/local/go/src/go/ast/walk.go:254 +0x1212
go/ast.walkStmtList(0x1473980, 0xc4200841e0, 0xc4205bc780, 0x5, 0x8)
	/usr/local/go/src/go/ast/walk.go:32 +0x81
go/ast.Walk(0x1473980, 0xc4200841e0, 0x1474dc0, 0xc42023b560)
	/usr/local/go/src/go/ast/walk.go:224 +0x1b71
go/ast.Walk(0x1473980, 0xc4200841e0, 0x1475240, 0xc42023b590)
	/usr/local/go/src/go/ast/walk.go:344 +0xd83
go/ast.walkDeclList(0x1473980, 0xc4200841e0, 0xc42046d400, 0x2b, 0x40)
	/usr/local/go/src/go/ast/walk.go:38 +0x81
go/ast.Walk(0x1473980, 0xc4200841e0, 0x14751c0, 0xc4205bcf80)
	/usr/local/go/src/go/ast/walk.go:353 +0x266f
github.com/GoASTScanner/gas.(*Analyzer).Process(0xc4200841e0, 0xc42004a8d0, 0x1, 0x1, 0x1, 0xc42004a8d0)
	/Users/browne/go/src/github.com/GoASTScanner/gas/analyzer.go:141 +0x57d
main.main()
	/Users/browne/go/src/github.com/GoASTScanner/gas/cmd/gas/main.go:224 +0x459
```

To reproduce:
```
go get golang.org/x/crypto/ssh
cd /Users/browne/workspace/go/src/golang.org
gas x/crypto/acme/autocert/
```

Hello, 
I also have such issue, but on another file - selected from a Hyperledger project. 
It's definitely related to G402
```
go version
go version go1.9.4 linux/amd64

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x4a3333]

goroutine 1 [running]:
panic(0x6b6a60, 0xc42000e0d0)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/GoASTScanner/gas/rules.(*insecureConfigTLS).Match(0xc42000b1a0, 0x84eca0, 0xc42012ad40, 0xc42012a3c0, 0x1, 0x84ee00, 0xc420653048)
        /home/peyo/test/hyper/projects/src/src/github.com/GoASTScanner/gas/rules/tls.go:111 +0x83
github.com/GoASTScanner/gas.(*Analyzer).Visit(0xc42004e140, 0x84eca0, 0xc42012ad40, 0x84d920, 0xc42004e140)
        /home/peyo/test/hyper/projects/src/src/github.com/GoASTScanner/gas/analyzer.go:171 +0x1e2
go/ast.Walk(0x84d920, 0xc42004e140, 0x84eca0, 0xc42012ad40)
        /usr/local/go/src/go/ast/walk.go:52 +0x63
go/ast.Walk(0x84d920, 0xc42004e140, 0x84eee0, 0xc4203ba2a0)
        /usr/local/go/src/go/ast/walk.go:151 +0x43e
go/ast.walkExprList(0x84d920, 0xc42004e140, 0xc42012ad80, 0x3, 0x4)
        /usr/local/go/src/go/ast/walk.go:26 +0x89
go/ast.Walk(0x84d920, 0xc42004e140, 0x84eca0, 0xc42012adc0)
        /usr/local/go/src/go/ast/walk.go:104 +0x551
go/ast.walkExprList(0x84d920, 0xc42004e140, 0xc420082d80, 0x5, 0x8)
        /usr/local/go/src/go/ast/walk.go:26 +0x89
go/ast.Walk(0x84d920, 0xc42004e140, 0x84eca0, 0xc42012b140)
        /usr/local/go/src/go/ast/walk.go:104 +0x551
go/ast.walkExprList(0x84d920, 0xc42004e140, 0xc42012b480, 0x3, 0x4)
        /usr/local/go/src/go/ast/walk.go:26 +0x89
go/ast.Walk(0x84d920, 0xc42004e140, 0x84eba0, 0xc42012b4c0)
        /usr/local/go/src/go/ast/walk.go:137 +0x1233
go/ast.walkExprList(0x84d920, 0xc42004e140, 0xc420413920, 0x1, 0x1)
        /usr/local/go/src/go/ast/walk.go:26 +0x89
go/ast.Walk(0x84d920, 0xc42004e140, 0x84eaa0, 0xc42012b500)
        /usr/local/go/src/go/ast/walk.go:207 +0x2479
go/ast.walkStmtList(0x84d920, 0xc42004e140, 0xc42012b540, 0x3, 0x4)
        /usr/local/go/src/go/ast/walk.go:32 +0x89
go/ast.Walk(0x84d920, 0xc42004e140, 0x84eb60, 0xc4203baae0)
        /usr/local/go/src/go/ast/walk.go:224 +0x1d48
go/ast.Walk(0x84d920, 0xc42004e140, 0x7faeff464b10, 0xc4203bab10)
        /usr/local/go/src/go/ast/walk.go:344 +0xf03
go/ast.walkDeclList(0x84d920, 0xc42004e140, 0xc420082e80, 0x8, 0x8)
        /usr/local/go/src/go/ast/walk.go:38 +0x89
go/ast.Walk(0x84d920, 0xc42004e140, 0x84eda0, 0xc420082f00)
        /usr/local/go/src/go/ast/walk.go:353 +0x29e4
github.com/GoASTScanner/gas.(*Analyzer).Process(0xc42004e140, 0xc4204129d0, 0x1, 0x1, 0x1, 0xc4204129d0)
        /home/peyo/test/hyper/projects/src/src/github.com/GoASTScanner/gas/analyzer.go:141 +0xb00
main.main()
        /home/peyo/test/hyper/projects/src/src/github.com/GoASTScanner/gas/cmd/gas/main.go:224 +0x437



```
Attached file is renamed to .txt

[fsm.txt](https://github.com/GoASTScanner/gas/files/1751630/fsm.txt)

@peyoh Please could you check if the panic still occur with this fix? It would be great if you could point me to the package from where the attached code originates.

Somehow, I cannot reproduce it on my machine:
```
gas ./...
[gas] 2018/02/23 14:14:39 including rules: default
[gas] 2018/02/23 14:14:39 excluding rules: default
[gas] 2018/02/23 14:14:39 Searching directory: /home/cosmin/go/src/github.com/cosmincojocar/gas
/home/cosmin/go/src/github.com/cosmincojocar/gas/fsm.go:55:2: undeclared name: logger
/home/cosmin/go/src/github.com/cosmincojocar/gas/fsm.go:59:2: undeclared name: logger
/home/cosmin/go/src/github.com/cosmincojocar/gas/fsm.go:63:2: undeclared name: logger
/home/cosmin/go/src/github.com/cosmincojocar/gas/fsm.go:67:2: undeclared name: logger
/home/cosmin/go/src/github.com/cosmincojocar/gas/fsm.go:71:2: undeclared name: logger
[gas] 2018/02/23 14:14:40 Checking package: core
[gas] 2018/02/23 14:14:40 Checking file: /home/cosmin/go/src/github.com/cosmincojocar/gas/fsm.go
Results:

Summary:
   Files: 1
   Lines: 72
   Nosec: 0
  Issues: 0
```

Hello, 
the issue is still here. 

#uname -na
Linux  3.10.0-693.17.1.el7.x86_64 #1 SMP Thu Jan 25 20:13:58 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
#go version
#go version go1.9.4 linux/amd64
#git rev-parse HEAD edb362fc9d80dd292e27f32a8070494d52d66ed9
#git checkout
#go build -o gas ./cmd/gas/
# cd /home/test/fabric/core
# gas -include=G402 ./...
```
[gas] 2018/02/27 14:46:34 Checking file: /home/test/hyper/fabric/core/fsm.go
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x66cfbc]

goroutine 1 [running]:
github.com/GoASTScanner/gas/rules.(*insecureConfigTLS).Match(0xc42000eff0, 0x87a020, 0xc4202bfc40, 0xc420054400, 0x879d01, 0xc42e2e6a08, 0x513b4e)
        /home/peyo/test/hyper/projects/src/src/github.com/GoASTScanner/gas/rules/tls.go:111 +0x7c
github.com/GoASTScanner/gas.(*Analyzer).Visit(0xc4200841e0, 0x87a020, 0xc4202bfc40, 0x8789a0, 0xc4200841e0)
        /home/peyo/test/hyper/projects/src/src/github.com/GoASTScanner/gas/analyzer.go:171 +0x1c5
go/ast.Walk(0x8789a0, 0xc4200841e0, 0x87a020, 0xc4202bfc40)
        /usr/local/go/src/go/ast/walk.go:52 +0x66
go/ast.Walk(0x8789a0, 0xc4200841e0, 0x87a560, 0xc42023a5d0)
        /usr/local/go/src/go/ast/walk.go:151 +0x3b5
go/ast.walkExprList(0x8789a0, 0xc4200841e0, 0xc4202bfc80, 0x3, 0x4)
        /usr/local/go/src/go/ast/walk.go:26 +0x81
go/ast.Walk(0x8789a0, 0xc4200841e0, 0x87a020, 0xc4202bfcc0)
        /usr/local/go/src/go/ast/walk.go:104 +0x43e
go/ast.walkExprList(0x8789a0, 0xc4200841e0, 0xc420ce3200, 0x5, 0x8)
        /usr/local/go/src/go/ast/walk.go:26 +0x81
go/ast.Walk(0x8789a0, 0xc4200841e0, 0x87a020, 0xc420272040)
        /usr/local/go/src/go/ast/walk.go:104 +0x43e
go/ast.walkExprList(0x8789a0, 0xc4200841e0, 0xc420272380, 0x3, 0x4)
        /usr/local/go/src/go/ast/walk.go:26 +0x81
go/ast.Walk(0x8789a0, 0xc4200841e0, 0x879ea0, 0xc4202723c0)
        /usr/local/go/src/go/ast/walk.go:137 +0x1082
go/ast.walkExprList(0x8789a0, 0xc4200841e0, 0xc420cc6e50, 0x1, 0x1)
        /usr/local/go/src/go/ast/walk.go:26 +0x81
go/ast.Walk(0x8789a0, 0xc4200841e0, 0x879ca0, 0xc420272400)
        /usr/local/go/src/go/ast/walk.go:207 +0x211f
go/ast.walkStmtList(0x8789a0, 0xc4200841e0, 0xc420272440, 0x3, 0x4)
        /usr/local/go/src/go/ast/walk.go:32 +0x81
go/ast.Walk(0x8789a0, 0xc4200841e0, 0x879e20, 0xc42023ae10)
        /usr/local/go/src/go/ast/walk.go:224 +0x1b71
go/ast.Walk(0x8789a0, 0xc4200841e0, 0x87a2a0, 0xc42023ae40)
        /usr/local/go/src/go/ast/walk.go:344 +0xd83
go/ast.walkDeclList(0x8789a0, 0xc4200841e0, 0xc420ce3300, 0x8, 0x8)
        /usr/local/go/src/go/ast/walk.go:38 +0x81
go/ast.Walk(0x8789a0, 0xc4200841e0, 0x87a220, 0xc420ce3380)
        /usr/local/go/src/go/ast/walk.go:353 +0x266f
github.com/GoASTScanner/gas.(*Analyzer).Process(0xc4200841e0, 0xc420d50800, 0x5c, 0x80, 0x41, 0xc420d50800)
        /home/peyo/test/hyper/projects/src/src/github.com/GoASTScanner/gas/analyzer.go:141 +0x57d
main.main()
        /home/peyo/test/hyper/projects/src/src/github.com/GoASTScanner/gas/cmd/gas/main.go:224 +0x459

```
#

--- a/rules/tls.go
+++ b/rules/tls.go
@@ -108,12 +108,15 @@ func (t *insecureConfigTLS) processTLSConfVal(n *ast.KeyValueExpr, c *gas.Contex
 }
 
 func (t *insecureConfigTLS) Match(n ast.Node, c *gas.Context) (*gas.Issue, error) {
-	if complit, ok := n.(*ast.CompositeLit); ok && complit.Type != nil && c.Info.TypeOf(complit.Type).String() == t.requiredType {
-		for _, elt := range complit.Elts {
-			if kve, ok := elt.(*ast.KeyValueExpr); ok {
-				issue := t.processTLSConfVal(kve, c)
-				if issue != nil {
-					return issue, nil
+	if complit, ok := n.(*ast.CompositeLit); ok && complit.Type != nil {
+		actualType := c.Info.TypeOf(complit.Type)
+		if actualType != nil && actualType.String() == t.requiredType {
+			for _, elt := range complit.Elts {
+				if kve, ok := elt.(*ast.KeyValueExpr); ok {
+					issue := t.processTLSConfVal(kve, c)
+					if issue != nil {
+						return issue, nil
+					}
 				}
 			}
 		}

